python/ray/train/callbacks/profile.py

```diff
@@ -0,0 +1,113 @@
+import logging
+import tempfile
+from pathlib import Path
+from typing import List, Dict, Optional, Union
+
+from torch.profiler import profile
+
+from ray import train
+from ray.train import TrainingCallback
+from ray.train.callbacks.logging import TrainCallbackLogdirManager
+from ray.train.callbacks.results_preprocessors import IndexedResultsPreprocessor
+from ray.train.constants import PYTORCH_PROFILER_KEY
+
+logger = logging.getLogger(__name__)
+
+WORKER_TRACE_DIR_NAME = "pytorch_profiler_worker_traces"
+DRIVER_TRACE_DIR_NAME = "pytorch_profiler"
+
+
+class TorchWorkerProfiler():
+    """Utility class for running PyTorch Profiler on a Train worker.
+
+     Args:
+         trace_dir (Optional[str]): The directory to store traces on the
+            worker node. If ``None``, this will use a default temporary dir.
+    """
+
+    def __init__(self, trace_dir: Optional[str] = None):
+        trace_dir = trace_dir or Path(
+            tempfile.gettempdir()).joinpath(WORKER_TRACE_DIR_NAME)
+        self.trace_dir = Path(trace_dir)
+        self.trace_dir.mkdir(parents=True, exist_ok=True)
+        # Accumulated traces.
+        self.profiler_trace_filenames = []
+
+    def trace_handler(self, p: profile):
+        """A stateful PyTorch Profiler trace handler.
+
+        This will the export chrome trace to a file on disk.
+
+        These exported traces can then be fetched by calling
+        ``get_and_clear_profile_traces``.
+
+        Args:
+            p (profile): A PyTorch Profiler profile.
+        """
+        trace_filename = \
+            f"worker_{train.world_rank()}_epoch_{p.step_num}.pt.trace.json"
+        trace_path = self.trace_dir.joinpath(trace_filename)
+
+        logger.debug(f"Writing worker trace to {trace_path}.")
+        p.export_chrome_trace(str(trace_path))
+
+        self.profiler_trace_filenames.append(trace_filename)
+
+    def get_and_clear_profile_traces(self):
+        """Reads unread Profiler traces from this worker.
+
+        Returns:
+            The traces in a format consumable by
+            ``TorchTensorboardProfilerCallback``.
+        """
+
+        def get_trace(filename):
+            trace_path = self.trace_dir.joinpath(filename)
+            return trace_path.read_text()
+
+        traces = [(trace_filename, get_trace(trace_filename))
+                  for trace_filename in self.profiler_trace_filenames]
+
+        self.profiler_trace_files = []
+        return {PYTORCH_PROFILER_KEY: traces}
+
+
+class TorchTensorboardProfilerCallback(TrainingCallback):
```

Comment:
Is there a way to test this? I think even something that just checks if the files exist with the correct names should work.